src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -553,10 +565,69 @@ static void on_resolver_result_changed_locked(void* arg, grpc_error* error) {
       // Instantiate new LB policy.
       create_new_lb_policy_locked(chand, lb_policy_name.get(),
                                   &connectivity_state, &connectivity_error);
+      if (chand->channelz_channel != nullptr) {
+        char* str;
+        gpr_asprintf(&str, "Switched LB policy to %s", lb_policy_name.get());
+        trace_strings.push_back(str);
+      }
     }
     // Find service config.
     grpc_core::UniquePtr<char> service_config_json =
         get_service_config_from_resolver_result_locked(chand);
+    // Note: It's safe to use chand->info_service_config_json here without
+    // taking a lock on chand->info_mu, because this function is the
+    // only thing that modifies its value, and it can only be invoked
+    // once at any given time.
+    if (chand->channelz_channel != nullptr) {
+      if ((service_config_json == nullptr &&
+           chand->info_service_config_json != nullptr) ||
+          (service_config_json != nullptr &&
+           chand->info_service_config_json == nullptr) ||
+          (service_config_json != nullptr &&
+           chand->info_service_config_json != nullptr &&
```

Comment:
No need for this line.  If `service_config_json != nullptr`, then this must be true, since otherwise the conditions above would have been false.

Reply:
Done